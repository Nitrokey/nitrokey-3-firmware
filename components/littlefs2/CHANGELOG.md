# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.2.2] - 2021-03-20

### Changed
- Added `remove_dir_all_when`, allowing to filter "rm -rf <path>"

## [v0.2.1] - 2021-02-26

### Changed
- PathBuf::from errors on embedded nuls, and prevents ending
  with nuls
- get rid of ufmt (oversight in 0.2 release)
- get rid of dead code (oversight in 0.2 release)

## [v0.2.0] - 2021-02-02

### Changed

- [breaking-change] The version of the `generic-array` dependency has been
  bumped to v0.14.2 (now that `heapless` v0.6.0` is out).

## [v0.1.1] - 2021-02-11

### Fixed

- `std`-triggering regression
