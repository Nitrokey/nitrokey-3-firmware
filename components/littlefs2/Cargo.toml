[package]
name = "littlefs2"
description = "Idiomatic Rust API for littlefs"
version = "0.3.2"
authors = ["Nicolas Stalder <n@stalder.io>", "Brandon Edens <brandonedens@gmail.com>"]
edition = "2018"
license = "Apache-2.0 OR MIT"
readme = "README.md"
categories = ["embedded", "filesystem", "no-std"]
repository = "https://github.com/nickray/littlefs2"
resolver = "2"

[dependencies]
bitflags = "1"
cty = "0.2.1"
delog = "0.1.0"
generic-array = "0.14"
heapless = "0.7"

[dependencies.cstr_core]
default-features = false
version = "0.2"

[dependencies.littlefs2-sys]
version = "0.1.6"

[dependencies.serde]
version = "1"
default-features = false
features = ["derive"]
optional = true

[dev-dependencies]
ssmarshal = "1"
serde = { version = "1.0", default-features = false, features = ["derive"] }
# trybuild = "1"

[features]
default = ["dir-entry-path", "serde"]
# use experimental closure-based API
dir-entry-path = []
# enable assertions in backend C code
ll-assertions = ["littlefs2-sys/assertions"]
# enable trace in backend C code
ll-trace = ["littlefs2-sys/trace"]
c-stubs = []

log-all = []
log-none = []
log-info = []
log-debug = []
log-warn = []
log-error = []

# TODO: LFS_NAME_MAX (and maybe other sizes) are baked into the
# compiled C library. For instance, the `lfs_info` struct has a
# member `char name[LFS_NAME_MAX+1]`.
# This means that if we change `traits::Storage::FILENAME_MAX_PLUS_ONE`,
# we need to pass this on!
