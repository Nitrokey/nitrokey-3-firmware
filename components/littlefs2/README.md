<h1 align="center">littlefs2</h1>
<div align="center">
 <strong>
   Idiomatic Rust API for littlefs
 </strong>
</div>

<br />

<div align="center">
  <!-- Crates version -->
  <a href="https://crates.io/crates/littlefs2">
    <img src="https://img.shields.io/crates/v/littlefs2.svg?style=flat-square"
    alt="Crates.io version" />
  </a>
  <!-- API docs -->
  <a href="https://docs.rs/littlefs2">
    <img src="https://img.shields.io/badge/docs-latest-blue.svg?style=flat-square"
      alt="API docs" />
  </a>
  <!-- Continuous build -->
  <a href="https://github.com/nickray/littlefs2/actions?query=branch%3Amain">
    <img src="https://img.shields.io/github/workflow/status/nickray/littlefs2/CI/main?style=for-the-badge"
      alt="CI" height="20"/>
  </a>
</div>

## What is this?

Idiomatic Rust API for the [littlefs][littlefs] microcontroller filesystem by [Chris Haster][geky].

Number `2` refers to the on-disk format version, [supporting inline files, custom attributes and dynamic wear-leveling][release-notes-2].

We follow [`std::fs`][std-fs] as much as reasonable.

The low-level bindings are provided by the [littlefs2-sys][littlefs2-sys] library.

Upstream release: [v2.1.4][upstream-release]

[geky]: https://github.com/geky
[littlefs]: https://github.com/ARMmbed/littlefs
[release-notes-2]: https://github.com/ARMmbed/littlefs/releases/tag/v2.0.0
[std-fs]: https://doc.rust-lang.org/std/fs/index.html
[littlefs2-sys]: https://lib.rs/littlefs2-sys
[upstream-release]: https://github.com/ARMmbed/littlefs/releases/tag/v2.1.4

## `no_std`

This library is `no_std` compatible, but there are two gotchas.

- The dev-dependency `memchr` of `littlefs2-sys` has its `std` features activated. To prevent this, upgrade to at least Rust 1.51
  and add `resolver = "2"` in the consuming code's `[package]` section. This will be the default in Rust 2021 edition.

- At link time, `lfs.c` has a dependency on `strcpy`. When not linking to a `libc` with this symbol, activate the `c-stubs` feature
  to provide an implementation.

#### License

<sup>littlefs is licensed under [BSD-3-Clause](https://github.com/ARMmbed/littlefs/blob/master/LICENSE.md).</sup>
<sup>This API for littlefs is licensed under either of [Apache License, Version 2.0](LICENSE-APACHE) or [MIT License](LICENSE-MIT) at your option.</sup>
<sup>Previous bindings exist in the [rust-littlefs](https://github.com/brandonedens/rust-littlefs) repository, also dual-licensed under Apache-2.0 and MIT.</sup>
<br>
<sub>Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.</sub>
