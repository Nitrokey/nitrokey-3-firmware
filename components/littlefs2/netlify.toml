[build]
  # command = "curl https://sh.rustup.rs -sSf | sh -s -- -y && source $HOME/.cargo/env && sudo apt-get install llvm && cargo doc"
  command = "curl https://sh.rustup.rs -sSf | sh -s -- -y && source $HOME/.cargo/env && env && cargo doc"
  publish = "target/doc"

[[redirects]]
  from = "/"
  to = "/littlefs2"
