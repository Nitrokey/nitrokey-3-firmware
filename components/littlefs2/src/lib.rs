#![cfg_attr(not(test), no_std)]

/*!

[littlefs](https://github.com/ARMmbed/littlefs) is a filesystem for microcontrollers
written in C, that claims to be *fail-safe*:
- power-loss resilience, by virtue of copy-on-write guarantees
- bounded RAM/ROM, with stack-allocated buffers

Since [version 2](https://github.com/ARMmbed/littlefs/releases/tag/v2.0.0), it has
some nifty features such as:
- dynamic wear-leveling, including detection of bad Flash blocks
- custom user attributes
- inline files, avoiding block waste

For more background, see its [design notes](https://github.com/ARMmbed/littlefs/blob/master/DESIGN.md)
and the [specification](https://github.com/ARMmbed/littlefs/blob/master/SPEC.md) of its format.

### What is this?

This library, [`littlefs2`](https://lib.rs/littlefs2), offers an idiomatic Rust API for littlefs.

It follows the design of [`std::fs`](https://doc.rust-lang.org/std/fs/index.html) as much as reasonable,
and builds on the bindings [`littlefs2-sys`](https://lib.rs/littlefs2-sys).

Some complications arise due to the lack of const generics in Rust, we work around these
with the [`generic-array`](https://lib.rs/generic-array) library, and long for the day when
constants associated to traits will be treated as constants by the compiler.

Another complication is the fact that files (and directories) need to be closed before they go out of scope,
since the main littlefs state structure contains a linked list which would exhibit UB (undefined behaviour)
otherwise, see [issue #3](https://github.com/nickray/littlefs2/issues/3) and
[issue #5](https://github.com/nickray/littlefs2/issues/5). We choose *not* to call `close` in `drop` (as
`std::fs` does), since these operations could panic if for instance `littlefs` detects Flash corruption
(from which the application might otherwise recover).

For this reason, the various `File`-related `open` methods are marked as `unsafe`.
Instead, a closure-based API is offered (`open_and_then` and friends),
the same is done for `Filesystem::read_dir`. Under the hood, this API first calls the unsafe constructor,
then calls the user-supplied closure, and finally closes the object.

**FOLLOWING SECTION OUT-OF-DATE**

⯈ [**The best place to start reading the API docs is here**](fs/index.html). ⯇

### Usage

To use this library, implement `littlefs2::driver::Storage`.
The macro `ram_storage!` generates examples of this.

Roughly speaking, the [`Storage`](driver/trait.Storage.html) trait defines a block device in
terms of actual and `typenum` constants, and an implementation supplies methods to read, erase and write.

The filesystem and each open file need memory for state and caching, this has to be allocated
beforehand and passed to constructors.

### `no_std`

This library is `no_std` compatible, but there are two gotchas.

- The dev-dependency `memchr` of `littlefs2-sys` has its `std` features activated. To prevent this, upgrade to at least Rust 1.51
  and add `resolver = "2"` in the consuming code's `[package]` section. This will be the default in Rust 2021 edition.

- At link time, `lfs.c` has a dependency on `strcpy`. When not linking to a `libc` with this symbol, activate the `c-stubs` feature
  to provide an implementation.

### Design notes

All operations on the filesystem require passing a `&mut Storage`, which guarantees by Rust's
borrow checker that only one thread can manipulate the filesystem.
This design choice (as opposed to consuming the Storage, which would be less verbose) was made to
enable use of the underlying flash peripheral outside of the filesystem (the `Storage` can be
dropped and reconstructed).  For instance, one could setup an additional filesystem,
or handle some flash data manually.

As an experiment, we implemented [`ReadDirWith`](fs/struct.ReadDirWith.html). It converts a
[`ReadDir`](fs/struct.ReadDir.html) (which needs mutable references, and so is "not quite an iterator"
over the files of a directory), into a true iterator, by temporarily binding the mutable references.

Currying with lifetime gymnastics!

In the future, we may extend this approach to other operations, thus adding a secondary API layer.

<https://play.rust-lang.org/?edition=2018&gist=c86abf99fc87551cfe3136e398a45d19>

Separately, keeping track of the allocations is a chore, we hope that
[`Pin`](https://doc.rust-lang.org/core/pin/index.html) magic will help fix this.

### Example

```
# use littlefs2::fs::{Filesystem, File, OpenOptions};
# use littlefs2::io::prelude::*;
# use littlefs2::path::PathBuf;
#
# use littlefs2::{consts, ram_storage, driver, io::Result};
#
#
// example storage backend
ram_storage!(tiny);
let mut ram = Ram::default();
let mut storage = RamStorage::new(&mut ram);

// must format before first mount
Filesystem::format(&mut storage).unwrap();
// must allocate state statically before use
let mut alloc = Filesystem::allocate();
let mut fs = Filesystem::mount(&mut alloc, &mut storage).unwrap();

// may use common `OpenOptions`
let mut buf = [0u8; 11];
fs.open_file_with_options_and_then(
    |options| options.read(true).write(true).create(true),
    &PathBuf::from(b"example.txt"),
    |file| {
        file.write(b"Why is black smoke coming out?!")?;
        file.seek(SeekFrom::End(-24)).unwrap();
        assert_eq!(file.read(&mut buf)?, 11);
        Ok(())
    }
).unwrap();
assert_eq!(&buf, b"black smoke");
```
*/

/// Low-level bindings
pub use littlefs2_sys as ll;

#[macro_use]
extern crate delog;
generate_macros!();

/// cf. Macros documentation
#[macro_use]
pub mod macros;

#[cfg(feature = "c-stubs")]
mod c_stubs;

pub mod consts;
pub mod driver;

pub mod fs;
pub mod io;
pub mod path;

/// get information about the C backend
pub fn version() -> Version {
    Version {
        format: (ll::LFS_DISK_VERSION_MAJOR, ll::LFS_DISK_VERSION_MINOR),
        backend: (ll::LFS_VERSION_MAJOR, ll::LFS_VERSION_MINOR),
    }
}

/// Information about the C backend
#[derive(Clone,Copy,Debug)]
pub struct Version {
	/// On-disk format (currently: 2.0)
    pub format: (u32, u32),
	/// Backend release (currently: 2.1)
    pub backend: (u32, u32),
}

#[cfg(test)]
mod tests;
