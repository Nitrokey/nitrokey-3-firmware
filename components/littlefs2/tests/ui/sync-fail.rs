use littlefs2::{
    consts,
    fs::{File, Filesystem},
    io::{Error, Result, Write},
    ram_storage, driver,
};

ram_storage!(
    name=OtherRamStorage,
    backend=OtherRam,
    trait=driver::Storage,
    erase_value=0xff,
    read_size=1,
    write_size=32,
    cache_size_ty=consts::U32,
    block_size=256,
    block_count=512,
    lookaheadwords_size_ty=consts::U1,
    filename_max_plus_one_ty=consts::U256,
    path_max_plus_one_ty=consts::U256,
    result=Result,
);

ram_storage!(
    name=RamStorage,
    backend=Ram,
    trait=driver::Storage,
    erase_value=0xff,
    read_size=20*5,
    write_size=20*7,
    cache_size_ty=consts::U700,
    block_size=20*35,
    block_count=32,
    lookaheadwords_size_ty=consts::U1,
    filename_max_plus_one_ty=consts::U256,
    path_max_plus_one_ty=consts::U256,
    result=Result,
);

fn main() {
    let mut ram = Ram::default();
    let mut storage = RamStorage::new(&mut ram);
    let mut alloc = Filesystem::allocate();
    Filesystem::format(&mut storage).unwrap();
    let mut fs = Filesystem::mount(&mut alloc, &mut storage).unwrap();

    let mut other_ram = OtherRam::default();
    let mut other_storage = OtherRamStorage::new(&mut other_ram);
    let mut alloc = Filesystem::allocate();
    assert!(Filesystem::format(&mut other_storage).is_ok());
    let other_fs = Filesystem::mount(&mut alloc, &mut other_storage).unwrap();

    let mut alloc = File::allocate();
    // file does not exist yet, can't open for reading
    assert_eq!(
        File::open("/test_open.txt", &mut alloc, &mut fs, &mut storage)
            .map(drop)
            .unwrap_err(), // "real" contains_err is experimental
        Error::NoSuchEntry
    );

    fs.create_dir("/tmp", &mut storage).unwrap();

    // TODO: make previous allocation reusable
    let mut alloc = File::allocate();
    // can create new files
    let mut file = File::create("/tmp/test_open.txt", &mut alloc, &mut fs, &mut storage).unwrap();
    // can write to files
    assert!(file.write(&mut fs, &mut storage, &[0u8, 1, 2]).unwrap() == 3);
    // won't work: "expected `RamStorage`, found `RamStorageNormal`.
    file.sync(&mut other_fs, &mut other_storage).unwrap();
}
