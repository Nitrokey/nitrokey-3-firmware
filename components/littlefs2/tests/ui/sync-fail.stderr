error[E0308]: mismatched types
  --> $DIR/sync-fail.rs:71:15
   |
71 |     file.sync(&mut other_fs, &mut other_storage).unwrap();
   |               ^^^^^^^^^^^^^ expected struct `RamStorage`, found struct `OtherRamStorage`
   |
   = note: expected type `&mut littlefs2::fs::Filesystem<'_, RamStorage<'_>>`
              found type `&mut littlefs2::fs::Filesystem<'_, OtherRamStorage<'_>>`

error[E0308]: mismatched types
  --> $DIR/sync-fail.rs:71:30
   |
71 |     file.sync(&mut other_fs, &mut other_storage).unwrap();
   |                              ^^^^^^^^^^^^^^^^^^ expected struct `RamStorage`, found struct `OtherRamStorage`
   |
   = note: expected type `&mut RamStorage<'_>`
              found type `&mut OtherRamStorage<'_>`
